# Terragrunt can't pull private repositories with ssh on gh actions

**pputman-clabs** commented *Jun 9, 2022*

I have the log of the issue below.  I have one repository with my terragrunt configuration, calling another private repository in github with my terraform modules.
I have verified my ssh key works because I've used it with other github actions to pull down the private repository, and if I set that private repository to public, then I can run terragrunt and pull down the modules.

Also,  can run this terragrunt setup locally with the same ssh key and pull down the terraform modules.

Is this a problem with terragrunt, or github actions?  Seems like something weird when both of them are together, because separately each one can access the repo, its only when I use terragrunt on github actions that it can't.



with:
    tf_actions_version: latest
    tg_actions_version: latest
    tf_actions_subcommand: init
    tf_actions_working_dir: ./okta/dev/accounts/groups
    tf_actions_comment: true
    tf_actions_binary: terragrunt
    tf_actions_cli_credentials_hostname: app.terraform.io
    tf_actions_fmt_write: false
  env:
    OKTA_API_TOKEN_DEV: ***
    tf_version: latest
    tg_version: latest
    tf_actions_dev_groups_working_dir: ./okta/dev/accounts/groups
    tf_actions_dev_users_working_dir: ./okta/dev/accounts/users
    tf_actions_dev_user_profiles_working_dir: ./okta/dev/accounts/user-profiles
    tf_actions_dev_aws_app_working_dir: ./okta/dev/apps/aws
    tf_actions_dev_google_workspaces_app_working_dir: ./okta/dev/apps/google-workspaces
    tf_actions_dev_google_cloud_app_working_dir: ./okta/dev/apps/google-cloud
    SSH_AUTH_SOCK: /tmp/ssh-M2Ypm[3](https://github.com/celo-org/okta-live-config/runs/6115439556?check_suite_focus=true#step:7:3)DzkgvD/agent.1760
    SSH_AGENT_PID: 1761
    GITHUB_TOKEN: ***
/usr/bin/docker run --name b319c02[4](https://github.com/celo-org/okta-live-config/runs/6115439556?check_suite_focus=true#step:7:4)c7e374446a2b7141d6b[5](https://github.com/celo-org/okta-live-config/runs/6115439556?check_suite_focus=true#step:7:5)bd8[6](https://github.com/celo-org/okta-live-config/runs/6115439556?check_suite_focus=true#step:7:6)4cb_f42488 --label 294b31 --workdir /github/workspace --rm -e OKTA_API_TOKEN_DEV -e tf_version -e tg_version -e tf_actions_dev_groups_working_dir -e tf_actions_dev_users_working_dir -e tf_actions_dev_user_profiles_working_dir -e tf_actions_dev_aws_app_working_dir -e tf_actions_dev_google_workspaces_app_working_dir -e tf_actions_dev_google_cloud_app_working_dir -e SSH_AUTH_SOCK -e SSH_AGENT_PID -e GITHUB_TOKEN -e INPUT_TF_ACTIONS_VERSION -e INPUT_TG_ACTIONS_VERSION -e INPUT_TF_ACTIONS_SUBCOMMAND -e INPUT_TF_ACTIONS_WORKING_DIR -e INPUT_TF_ACTIONS_COMMENT -e INPUT_TF_ACTIONS_BINARY -e INPUT_TF_ACTIONS_CLI_CREDENTIALS_HOSTNAME -e INPUT_TF_ACTIONS_CLI_CREDENTIALS_TOKEN -e INPUT_TF_ACTIONS_FMT_WRITE -e HOME -e GITHUB_JOB -e GITHUB_REF -e GITHUB_SHA -e GITHUB_REPOSITORY -e GITHUB_REPOSITORY_OWNER -e GITHUB_RUN_ID -e GITHUB_RUN_NUMBER -e GITHUB_RETENTION_DAYS -e GITHUB_RUN_ATTEMPT -e GITHUB_ACTOR -e GITHUB_WORKFLOW -e GITHUB_HEAD_REF -e GITHUB_BASE_REF -e GITHUB_EVENT_NAME -e GITHUB_SERVER_URL -e GITHUB_API_URL -e GITHUB_GRAPHQL_URL -e GITHUB_REF_NAME -e GITHUB_REF_PROTECTED -e GITHUB_REF_TYPE -e GITHUB_WORKSPACE -e GITHUB_ACTION -e GITHUB_EVENT_PATH -e GITHUB_ACTION_REPOSITORY -e GITHUB_ACTION_REF -e GITHUB_PATH -e GITHUB_ENV -e GITHUB_STEP_SUMMARY -e RUNNER_OS -e RUNNER_ARCH -e RUNNER_NAME -e RUNNER_TOOL_CACHE -e RUNNER_TEMP -e RUNNER_WORKSPACE -e ACTIONS_RUNTIME_URL -e ACTIONS_RUNTIME_TOKEN -e ACTIONS_CACHE_URL -e GITHUB_ACTIONS=true -e CI=true -v "/var/run/docker.sock":"/var/run/docker.sock" -v "/home/runner/work/_temp/_github_home":"/github/home" -v "/home/runner/work/_temp/_github_workflow":"/github/workflow" -v "/home/runner/work/_temp/_runner_file_commands":"/github/file_commands" -v "/home/runner/work/okta-live-config/okta-live-config":"/github/workspace" 294b31:9c024c[7](https://github.com/celo-org/okta-live-config/runs/6115439556?check_suite_focus=true#step:7:7)e374446a2b7141d6b5bd[8](https://github.com/celo-org/okta-live-config/runs/6115439556?check_suite_focus=true#step:7:8)64cb
Checking the latest version of Terraform
Downloading Terraform v1.1.[9](https://github.com/celo-org/okta-live-config/runs/6115439556?check_suite_focus=true#step:7:9)
Successfully downloaded Terraform v1.1.9
Unzipping Terraform v1.1.9
Successfully unzipped Terraform v1.1.9
Checking the latest version of Terragrunt
Downloading Terragrunt v0.36.7
Successfully downloaded Terragrunt v0.36.7
Moving Terragrunt v0.36.7 to PATH
Successfully moved Terragrunt v0.36.7
init: info: initializing Terragrunt configuration in ./okta/dev/accounts/groups
init: error: failed to initialize Terragrunt configuration in ./okta/dev/accounts/groups
time=2022-04-21T16:02:41Z level=error msg=1 error occurred:
	* error downloading 'ssh://git@github.com/celo-org/tf-service-catalog.git': /usr/bin/git exited with [12](https://github.com/celo-org/okta-live-config/runs/6115439556?check_suite_focus=true#step:7:12)8: Cloning into '/github/workspace/okta/dev/accounts/groups/.terragrunt-cache/nzkLeP6grxGpGN1eoHw1_Nk0vVk/fLaSWSh_aJYGXkJFdH61Aul2BUI'...
Host key verification failed.
fatal: Could not read from remote repository.
Please make sure you have the correct access rights
and the repository exists.
time=2022-04-21T[16](https://github.com/celo-org/okta-live-config/runs/6115439556?check_suite_focus=true#step:7:16):02:[41](https://github.com/celo-org/okta-live-config/runs/6115439556?check_suite_focus=true#step:7:42)Z level=error msg=Unable to determine underlying exit code, so Terragrunt will exit with error code 1
<br />
***


**denis256** commented *Jun 10, 2022*

Hi,
looks like SSH keys aren't set in the Github actions container, it will work if will be explicitly executed clone `git clone git@github.com:...` in action?
May be helpful if the private key will be explicitly saved in `.ssh` directory in github action execution: `echo $SSH_KEY_TO_ACCESS_GITHUB_REPOS > ~/.ssh/id_rsa`
***

**pputman-clabs** commented *Jun 10, 2022*

Hi Denis,

  I originally thought that was the case too, but I'm able to connect to
the github runner and do the following from its terminal:

[image: image.png]


***@***.***:~/work/okta-live-config/okta-live-config$ ssh -Tv
***@***.***
OpenSSH_8.2p1 Ubuntu-4ubuntu0.5, OpenSSL 1.1.1f  31 Mar 2020
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 19: include /etc/ssh/ssh_config.d/*.conf
matched no files
debug1: /etc/ssh/ssh_config line 21: Applying options for *
debug1: Connecting to github.com [140.82.113.3] port 22.
debug1: Connection established.
debug1: identity file /home/runner/.ssh/id_rsa type -1
debug1: identity file /home/runner/.ssh/id_rsa-cert type -1
debug1: identity file /home/runner/.ssh/id_dsa type -1
debug1: identity file /home/runner/.ssh/id_dsa-cert type -1
debug1: identity file /home/runner/.ssh/id_ecdsa type -1
debug1: identity file /home/runner/.ssh/id_ecdsa-cert type -1
debug1: identity file /home/runner/.ssh/id_ecdsa_sk type -1
debug1: identity file /home/runner/.ssh/id_ecdsa_sk-cert type -1
debug1: identity file /home/runner/.ssh/id_ed25519 type -1
debug1: identity file /home/runner/.ssh/id_ed25519-cert type -1
debug1: identity file /home/runner/.ssh/id_ed25519_sk type -1
debug1: identity file /home/runner/.ssh/id_ed25519_sk-cert type -1
debug1: identity file /home/runner/.ssh/id_xmss type -1
debug1: identity file /home/runner/.ssh/id_xmss-cert type -1
debug1: Local version string SSH-2.0-OpenSSH_8.2p1 Ubuntu-4ubuntu0.5
debug1: Remote protocol version 2.0, remote software version babeld-dd60f759
debug1: no match: babeld-dd60f759
debug1: Authenticating to github.com:22 as 'git'
debug1: SSH2_MSG_KEXINIT sent
debug1: SSH2_MSG_KEXINIT received
debug1: kex: algorithm: curve25519-sha256
debug1: kex: host key algorithm: ecdsa-sha2-nistp256
debug1: kex: server->client cipher: ***@***.*** MAC:
<implicit> compression: none
debug1: kex: client->server cipher: ***@***.*** MAC:
<implicit> compression: none
debug1: expecting SSH2_MSG_KEX_ECDH_REPLY
debug1: Server host key: ecdsa-sha2-nistp256
SHA256:p2QAMXNIC1TJYWeIOttrVc98/R1BUFWu3/LiyKgUfQM
debug1: Host 'github.com' is known and matches the ECDSA host key.
debug1: Found key in /home/runner/.ssh/known_hosts:1
Warning: Permanently added the ECDSA host key for IP address '140.82.113.3'
to the list of known hosts.
debug1: rekey out after 134217728 blocks
debug1: SSH2_MSG_NEWKEYS sent
debug1: expecting SSH2_MSG_NEWKEYS
debug1: SSH2_MSG_NEWKEYS received
debug1: rekey in after 134217728 blocks
debug1: Will attempt key: /home/runner/.ssh/id_rsa
debug1: Will attempt key: /home/runner/.ssh/id_dsa
debug1: Will attempt key: /home/runner/.ssh/id_ecdsa
debug1: Will attempt key: /home/runner/.ssh/id_ecdsa_sk
debug1: Will attempt key: /home/runner/.ssh/id_ed25519
debug1: Will attempt key: /home/runner/.ssh/id_ed25519_sk
debug1: Will attempt key: /home/runner/.ssh/id_xmss
debug1: SSH2_MSG_EXT_INFO received
debug1: kex_input_ext_info: server-sig-algs=<
***@***.******@***.***,
***@***.***,
***@***.******@***.***
***@***.***,
***@***.******@***.***,
***@***.******@***.***,
***@***.***
,ssh-ed25519,ecdsa-sha2-nistp521,ecdsa-sha2-nistp384,ecdsa-sha2-nistp256,rsa-sha2-512,rsa-sha2-256,ssh-rsa>
debug1: SSH2_MSG_SERVICE_ACCEPT received
debug1: Authentications that can continue: publickey
debug1: Next authentication method: publickey
debug1: Trying private key: /home/runner/.ssh/id_rsa
debug1: Authentication succeeded (publickey).
Authenticated to github.com ([140.82.113.3]:22).
debug1: channel 0: new [client-session]
debug1: Entering interactive session.
debug1: pledge: network
debug1: client_input_global_request: rtype ***@***.***
want_reply 0
debug1: Sending environment.
debug1: Sending env LANG = C.UTF-8
debug1: client_input_channel_req: channel 0 rtype exit-status reply 0
Hi celo-org/tf-service-catalog! You've successfully authenticated, but
GitHub does not provide shell access.
debug1: channel 0: free: client-session, nchannels 1
Transferred: sent 2704, received 2348 bytes, in 0.2 seconds
Bytes per second: sent 17230.2, received 14961.7
debug1: Exit status 1


So the ssh key appears to be working properly, but terragrunt isn't using
it for some reason? Not sure how to test this further.

On Fri, Jun 10, 2022 at 3:34 AM Denis O ***@***.***> wrote:

> Hi,
> looks like SSH keys aren't set in the Github actions container, it will
> work if will be explicitly executed clone git clone ***@***.***:... in
> action?
> May be helpful if the private key will be explicitly saved in .ssh
> directory in github action execution: echo
> $SSH_KEY_TO_ACCESS_GITHUB_REPOS > ~/.ssh/id_rsa
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/gruntwork-io/terragrunt/issues/2139#issuecomment-1152117489>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AX2F4DWGJGOZAIG4CLT6HH3VOL43BANCNFSM5YKTQKZQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

***

**pputman-clabs** commented *Jun 13, 2022*

I figured this out, the terragrunt github actions is putting the init command in a container, I had to throw the key into that container volume
***

**asaf-upstream** commented *May 31, 2023*

Hi @pputman-clabs , cloud you plz elaborate on your solution?
I am facing the same issue, probably with  the same gh action..
Did you forked it and added the key  into the new repo and dockerfile?
***

